Fix ip family mismatch for routes

d-uzlov opened the pull request:
Related issue: https://github.com/networkservicemesh/integration-k8s-packet/issues/68

denis-tingaikin commented on internal/vppinit/vppinit.go:
```diff
@@ -149,28 +150,33 @@ func LinkToAfPacket(ctx context.Context, vppConn api.Connection, tunnelIP net.IP
 			},
 		},
 	}
 	for _, route := range routes {
+		logger := log.FromContext(ctx).
+			WithField("swIfIndex", swIfIndex).
+			WithField("nh.address", types.FromVppIPAddressUnion(ipRouteAddDel.Route.Paths[0].Nh.Address, route.Gw.To4() == nil)).
+			WithField("prefix", ipRouteAddDel.Route.Prefix).
+			WithField("isAdd", true).
+			WithField("vppapi", "IPRouteAddDel")
 		ipRouteAddDel.Route.Prefix = types.ToVppPrefix(route.Dst)
 		if route.Gw != nil {
+			routeIsIpv6 := route.Gw.To4() == nil
+			if tunnelIsIpv6 != routeIsIpv6 {
+				logger.Debug("skipped addr with mismatched ip version")
+				continue
+			}
```
```suggestion
	for _, route := range routes {
	        isIPv6 := route.Gw.To4() == nil
		logger := log.FromContext(ctx).
			WithField("swIfIndex", swIfIndex).
			WithField("nh.address", types.FromVppIPAddressUnion(ipRouteAddDel.Route.Paths[0].Nh.Address, route.Gw.To4() == nil)).
			WithField("prefix", ipRouteAddDel.Route.Prefix).
			WithField("isAdd", true).
			WithField("vppapi", "IPRouteAddDel").
			WithField("isIPv6", isIPv6)
		if route.Gw == nil {
		   logger.Debug("todo")
		   continue
		}
		if isTunnelIPv6 != isIPv6 {
			logger.Debug("skipped addr with mismatched ip version")
			continue
		}
```

denis-tingaikin commented on internal/vppinit/vppinit.go:
```diff
@@ -149,28 +150,33 @@ func LinkToAfPacket(ctx context.Context, vppConn api.Connection, tunnelIP net.IP
 		},
 	}
 	for _, route := range routes {
+		logger := log.FromContext(ctx).
+			WithField("swIfIndex", swIfIndex).
+			WithField("nh.address", types.FromVppIPAddressUnion(ipRouteAddDel.Route.Paths[0].Nh.Address, route.Gw.To4() == nil)).
+			WithField("prefix", ipRouteAddDel.Route.Prefix).
+			WithField("isAdd", true).
+			WithField("vppapi", "IPRouteAddDel")
```
Please correct logging. We do not do IPRouteAddDel

denis-tingaikin commented on internal/vppinit/vppinit.go:
```diff
@@ -149,28 +150,33 @@ func LinkToAfPacket(ctx context.Context, vppConn api.Connection, tunnelIP net.IP
+			WithField("isAdd", true).
+			WithField("vppapi", "IPRouteAddDel")
 		ipRouteAddDel.Route.Prefix = types.ToVppPrefix(route.Dst)
 		if route.Gw != nil {
+			routeIsIpv6 := route.Gw.To4() == nil
+			if tunnelIsIpv6 != routeIsIpv6 {
+				logger.Debug("skipped addr with mismatched ip version")
+				continue
+			}
 			ipRouteAddDel.Route.Paths[0].Nh.Address = types.ToVppAddress(route.Gw).Un
```
Lets use two separate conds for simplifying

```
if !cond1 {
 continue
}
if !cond2 {
 continue
}
```

denis-tingaikin commented on internal/vppinit/vppinit.go:
```diff
@@ -149,28 +150,33 @@ func LinkToAfPacket(ctx context.Context, vppConn api.Connection, tunnelIP net.IP
-			WithField("isAdd", true).
-			WithField("duration", time.Since(now)).
-			WithField("vppapi", "IPRouteAddDel").Debug("completed")
+		logger.WithField("duration", time.Since(now)).Debug("completed")
```
Please add operation info...

denis-tingaikin replied:
Oops, this comment is invalid. I read it so quickly.

denis-tingaikin commented:
@edwarnicke This makes the packet green. It is tested by @d-uzlov .

Do you mind if we merge this?

edwarnicke commented:
@d-uzlov Could you explain the root cause this is fixing?  I can see what you are doing (leaving out routes that don't match the same address family as the tunnel IP).  I'm just super curious how adding other routes caused the issues we were seeing.

d-uzlov commented:
The root cause was that we override the route we need with a second route to ipv6.
We saw that strange `38.4.19.128` address due to ip version mismatch, which turned out to be `2604:1380:0:2c00::2` after I fixed the mismatch. However, after I added the fix, this ipv6 address still was in the place where `136.144.51.108` needs to be in the `vppctl show ip fib` output.
Judging by `ip route` output, we don't have corresponding routes for this ipv6 address. Which, probably, makes sense, since we wanted a tunnel with ipv4 address.

I'm not sure if this issue was linked to interfaces have multiple ipv4 addresses, as you mentioned. It was caused by having several addresses, but second address that overrode the correct route was ipv6.

edwarnicke commented:
Ah... OK... that makes lots of sense.

Question: If the route from the kernel has Gw of `2604:1380:0:2c00::2` ... does the route from the kernel have a Dst of `0.0.0.0/0` ?  I would have expected it to have a Dst of `::/0` ...

d-uzlov commented:
Both `136.144.51.108` and `2604:1380:0:2c00::2` routes have `Dst == nil`.

d-uzlov commented:
I have a feeling they should have had `0.0.0.0/0` and `::/0`, but somewhere serialization happened and stripped it.

edwarnicke commented:
Ah... OK... so here's a better solution.

If Dst == nil, we want to set the prefix to the prefix that matches the address family of Gw.

So if Dst == nil, and Gw is IPv4 - 0.0.0.0/0
If Dst == nil, and Gw is IPv6 ::/0

Make sense?  Could you try that solution?

d-uzlov commented:
Yeah, that would probably be better. I'll write here about results after I test this.

d-uzlov commented:
Your solution seems to be working perfectly
